projects/packages/image-cdn/src/class-image-cdn-core.php

```diff
@@ -0,0 +1,390 @@
+			$photon_url = add_query_arg( array( 'ssl' => 1 ), $photon_url );
+		}
+
+		return self::cdn_url_scheme( $photon_url, $scheme );
+	}
+
+	/**
+	 * Parses WP.com-hosted image args to replicate the crop.
+	 *
+	 * @param mixed  $args Args set during Photon's processing.
+	 * @param string $image_url URL of the image.
+	 * @return array|string Args for Photon to use for the URL.
+	 */
+	public static function parse_wpcom_query_args( $args, $image_url ) {
+		$parsed_url = wp_parse_url( $image_url );
+
+		if ( ! $parsed_url ) {
+			return $args;
+		}
+
+		$image_url_parts = wp_parse_args(
+			$parsed_url,
+			array(
+				'host'  => '',
+				'query' => '',
+			)
+		);
+
+		if ( '.files.wordpress.com' !== substr( $image_url_parts['host'], -20 ) ) {
+			return $args;
+		}
+
+		if ( empty( $image_url_parts['query'] ) ) {
+			return $args;
+		}
+
+		$wpcom_args = wp_parse_args( $image_url_parts['query'] );
+
+		if ( empty( $wpcom_args['w'] ) || empty( $wpcom_args['h'] ) ) {
+			return $args;
+		}
+
+		// Keep the crop by using "resize".
+		if ( ! empty( $wpcom_args['crop'] ) ) {
+			if ( is_array( $args ) ) {
+				$args = array_merge( array( 'resize' => array( $wpcom_args['w'], $wpcom_args['h'] ) ), $args );
+			} else {
+				$args = 'resize=' . rawurlencode( absint( $wpcom_args['w'] ) . ',' . absint( $wpcom_args['h'] ) ) . '&' . $args;
+			}
+		} elseif ( is_array( $args ) ) {
+			$args = array_merge( array( 'fit' => array( $wpcom_args['w'], $wpcom_args['h'] ) ), $args );
+		} else {
+			$args = 'fit=' . rawurlencode( absint( $wpcom_args['w'] ) . ',' . absint( $wpcom_args['h'] ) ) . '&' . $args;
+		}
+
+		return $args;
+	}
+
+	/**
+	 * Sets the scheme for a URL
+	 *
+	 * @param string $url URL to set scheme.
+	 * @param string $scheme Scheme to use. Accepts http, https, network_path.
+	 *
+	 * @return string URL.
+	 */
+	public static function cdn_url_scheme( $url, $scheme ) {
+		if ( ! in_array( $scheme, array( 'http', 'https', 'network_path' ), true ) ) {
+			if ( preg_match( '#^(https?:)?//#', $url ) ) {
+				return $url;
+			}
+
+			$scheme = 'http';
+		}
+
+		if ( 'network_path' === $scheme ) {
+			$scheme_slashes = '//';
+		} else {
+			$scheme_slashes = "$scheme://";
+		}
+
+		return preg_replace( '#^([a-z:]+)?//#i', $scheme_slashes, $url );
+	}
+
+	/**
+	 * Check to skip Photon for a known domain that shouldn't be Photonized.
+	 *
+	 * @param bool   $skip If the image should be skipped by Photon.
+	 * @param string $image_url URL of the image.
+	 *
+	 * @return bool Should the image be skipped by Photon.
+	 */
+	public static function banned_domains( $skip, $image_url ) {
+		$banned_host_patterns = array(
+			'/^chart\.googleapis\.com$/',
+			'/^chart\.apis\.google\.com$/',
+			'/^graph\.facebook\.com$/',
+			'/\.fbcdn\.net$/',
+			'/\.paypalobjects\.com$/',
+			'/\.dropbox\.com$/',
+			'/\.cdninstagram\.com$/',
+			'/^(commons|upload)\.wikimedia\.org$/',
+			'/\.wikipedia\.org$/',
```

Comment:
You know the Wiki infra a lot more than me, so I'll defer to you on that :)

> I wonder if it would be useful to match the path against /wiki/(?:File|Image):

This makes sense to me. it seems it would catch all the examples you mentioned; I don't know if there are any other URL formats that we need to be catching.

Reply:
The main other "format" we might care about catching is that the "File" and "Image" strings could be localized (and in any case). And while `/wiki/` should work for Wikimedia and Fandom, there's nothing requiring it. Even some Wikimedia sites might use something different in some cases, like https://zh.wikipedia.org/zh-tw/File:WordPress_logo.svg.